src/daft-core/src/array/ops/utf8.rs

```diff
@@ -378,6 +414,162 @@ impl Utf8Array {
+                        if regex {
+                            let regex_iter = pattern_arrow.iter().map(|pat| pat.map(regex::Regex::new));
+                            regex_replace(self_arrow.iter(), regex_iter, replacement_iter, self.name())
+                        } else {
+                            replace_on_literal(self_arrow.iter(), pattern_arrow.iter(), replacement_iter, self.name())
+                        }
+                    }
+                }
+            }
+            (1,1,replacement_len) => {
+                let self_scalar_value = self.get(0);
+                let pattern_scalar_value = pattern.get(0);
+                match (self_scalar_value, pattern_scalar_value) {
+                    (None, _) | (_, None) => Ok(Utf8Array::full_null(
+                        self.name(),
+                        self.data_type(),
+                        replacement_len,
+                    )),
+                    (Some(self_v), Some(pattern_v)) => {
+                        let arr_iter = std::iter::repeat(Some(self_v)).take(replacement_len);
+                        if regex {
+                            let re = Some(regex::Regex::new(pattern_v));
+                            let regex_iter = std::iter::repeat(re).take(replacement_len);
+                            regex_replace(arr_iter, regex_iter, replacement_arrow.iter(), self.name())
+                        } else {
+                            let pattern_iter = std::iter::repeat(Some(pattern_v)).take(replacement_len);
+                            replace_on_literal(arr_iter, pattern_iter, replacement_arrow.iter(), self.name())
+                        }
+                    }
+                }
+            },
+            (1,pattern_len,1) => {
+                let self_scalar_value = self.get(0);
+                let replacement_scalar_value = replacement.get(0);
+                match (self_scalar_value, replacement_scalar_value) {
+                    (None, _) | (_, None) => Ok(Utf8Array::full_null(
+                        self.name(),
+                        self.data_type(),
+                        pattern_len,
+                    )),
+                    (Some(self_v), Some(replacement_v)) => {
+                        let arr_iter = std::iter::repeat(Some(self_v)).take(pattern_len);
+                        let replacement_iter = std::iter::repeat(Some(replacement_v)).take(pattern_len);
+                        if regex {
+                            let regex_iter = pattern_arrow.iter().map(|pat| pat.map(regex::Regex::new));
+                            regex_replace(arr_iter, regex_iter, replacement_iter, self.name())
+                        } else {
+                            let pattern_iter = pattern_arrow.iter();
+                            replace_on_literal(arr_iter, pattern_iter, replacement_iter, self.name())
+                        }
+                    }
+                }
+            },
+            (self_len,1,1) => {
```

Comment:
Pretty crazy that we have so many match cases. I wonder if there's a better way to perform this.

Perhaps something like this at least?

```
self_iter, pattern_iter, replacement_iter = ...;
if regex {
    regex_replace(...)
} else {
    replace_on_literal(...)
}
```